onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -535,42 +535,23 @@ Status DP4AMatMulQuantizeProgram::GenerateShaderCode(ShaderHelper& shader) const
   shader.AddOutput("output", ShaderUsage::UseUniform);
   shader.AddOutput("scales", ShaderUsage::UseUniform);
 
-  shader.AdditionalImplementation() << R"ADDNL_FN(
-    var<workgroup> max_values : array<input_a_element_t, 4>;
- )ADDNL_FN";
-
   shader.MainFunctionBody() << R"MAIN_FN(
-  var local_a = input_a[global_idx];
-  var max_val = subgroupMax(abs(local_a));
-  var max_temp = max(max_val.xy, max_val.zw);
-  var scale = max(max_temp[0], max_temp[1]);
-  if (local_idx % sg_size == 0) {
-    max_values[local_idx / sg_size] = scale;
-  }
-  workgroupBarrier();
-
-  if (sg_size == 8)
+  var local_a : array<vec4<input_a_element_t>, 32>;
+  var max_value:vec4<input_a_element_t> = vec4<input_a_element_t>(0);
+  for (var idx:u32=0;idx<32;idx+=1)
   {
-    scale = max(max_values[0], max_values[1]);
-    scale = max(scale, max_values[2]);
-    scale = max(scale, max_values[3]);
+    local_a[idx] = input_a[workgroup_id.x*32 + idx];
```

Comment:
Please use `workgroup_idx` instead of `workgroup_id.x` in case workgroups is recalculated when one dimension is out of limitation. For example, when inputs > 1000, I see the result is not correct.

Reply:
workgroup_id.x and workgroup_idx  are exactly the same. workgroup_idx  is an ORT concept something ORT adds to the shader, as a prefix.

  let workgroup_idx = workgroup_id.x;